### PR TITLE
keep unsafe .. unsafe

### LINKED ISSUE #23734: !unsafe broken in Ansible 2.3.0.0?
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

jinja2 (?)

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /ansible/ansible.cfg
  configured module search path = [u'/ansible/library']
  python version = 2.7.12 (default, Jun 29 2016, 08:57:23) [GCC 5.3.0]

```

##### OS / ENVIRONMENT

Docker Alpine container running on Ubuntu

##### SUMMARY

Templating out...
```
!unsafe |
            ALERT InstanceDown
            IF up == 0
            FOR 5m
            LABELS { severity = "page" }
            ANNOTATIONS {
              summary = "Instance {{ $labels.instance }} down",
              description = "{{ $labels.instance }} of job {{ $labels.job }} has been down for more than 5 minutes.",
            }
```
Should result in the `{{ $labels.instance }}` and  `{{ $labels.job }}` remaining unchanged (i.e. not interpolated) but infact Ansible is now interpolating them again.

##### STEPS TO REPRODUCE

```yaml

```

##### EXPECTED RESULTS

For variables in unsafe strings to remain un-interpolated.

##### ACTUAL RESULTS

The string was interpolated

### PR DESCRIPTION
**Pull request by bcoca**
##### SUMMARY
fixes #23734, which was broken in previous fix that allowed non string types to be templated

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

templating

##### ANSIBLE VERSION

```
2.3/2.4
```

```

### CODE REVIEW

File: test/integration/targets/win_path/tasks/main.yml [hunk 1]
```diff
@@ -23,8 +23,8 @@
   assert:
     that:
     - item.0 | changed
-    - item.0.path_value == "C:\\{{ item.0.item }}Path"
-    - item.1.stdout_lines[0] == 'C:\\{{ item.0.item }}Path'
+    - item.0.path_value == "C:\\" + item.0.item + "Path"
```
Comment (alikins) [inline test/integration/targets/win_path/tasks/main.yml:26]: Is this and the test/integration/targets/copy/tasks/main.yml changes intentional? They seem unrelated.

File: lib/ansible/template/__init__.py [hunk 2]
```diff
@@ -482,6 +499,16 @@ def template(self, variable, convert_bare=False, preserve_trailing_newlines=True
...
 
+    def is_template(self, data):
+        ''' lets us know if data has a template'''
+        try:
+            new = self.template(data)
+        except UndefinedError:
+            return True
+        except:
+            return False
+        return (new != data)
```
Comment (alikins) [inline lib/ansible/template/__init__.py]: I added a couple of test cases to https://github.com/bcoca/ansible/compare/unsafe_unsafe_again...alikins:test_is_template that include passing in a tuple as 'data', and that seems to cause is_template to always return True

```
_______________________________________________________________________ TestTemplarIsTemplate.test_is_template_just_a_tuple _______________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_just_a_tuple>

    def test_is_template_just_a_tuple(self):
        some_obj = ('foo', 'bar', 1, 3, 5, None)
        res = self.templar.is_template(some_obj)
>       self.assertFalse(res, 'is_template(%s) return should be False' % repr(some_obj))
E       AssertionError: is_template(('foo', 'bar', 1, 3, 5, None)) return should be False

test/units/template/test_templar.py:347: AssertionError
_________________________________________________________________ TestTemplarIsTemplate.test_is_template_just_a_tuple_with_false __________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_just_a_tuple_with_false>

    def test_is_template_just_a_tuple_with_false(self):
        some_obj = (False, 'foo', 'bar', 1, 3, 5, None)
        res = self.templar.is_template(some_obj)
>       self.assertFalse(res, 'is_template(%s) return should be False' % repr(some_obj))
E       AssertionError: is_template((False, 'foo', 'bar', 1, 3, 5, None)) return should be False

test/units/template/test_templar.py:352: AssertionError
__________________________________________________________________ TestTemplarIsTemplate.test_is_template_just_a_tuple_with_true __________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_just_a_tuple_with_true>

    def test_is_template_just_a_tuple_with_true(self):
        some_obj = (True, 'foo', 'bar', 1, 3, 5, None)
        res = self.templar.is_template(some_obj)
>       self.assertFalse(res, 'is_template(%s) return should be False' % repr(some_obj))
E       AssertionError: is_template((True, 'foo', 'bar', 1, 3, 5, None)) return should be False

test/units/template/test_templar.py:357: AssertionError
________________________________________________________________ TestTemplarIsTemplate.test_is_template_tuple_with_unsafe_template ________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_tuple_with_unsafe_template>

    def test_is_template_tuple_with_unsafe_template(self):
        some_obj = ('foo', 'bar', wrap_var('{{ some_var }}'))
        res = self.templar.is_template(some_obj)
>       self.assertFalse(res, 'is_template(%s) return should be False' % repr(some_obj))
E       AssertionError: is_template(('foo', 'bar', u'{{ some_var }}')) return should be False

test/units/template/test_templar.py:367: AssertionError
```

File: lib/ansible/template/__init__.py [hunk 3]
```diff
@@ -482,6 +499,16 @@ def template(self, variable, convert_bare=False, preserve_trailing_newlines=True
             else:
                 return variable
 
+    def is_template(self, data):
+        ''' lets us know if data has a template'''
+        try:
+            new = self.template(data)
+        except UndefinedError:
+            return True
```
Comment (alikins) [inline lib/ansible/template/__init__.py]: I have a test case or two that fails somewhere in here, if the template string includes a '{%something%}'.
Not sure what the expected behavior is for that.

```
__________________________________________________________________________ TestTemplarIsTemplate.test_is_template_block ___________________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_block>

    def test_is_template_block(self):
        res = self.templar.is_template(u'1 2 {%what%} 3 4 {{ some_var }} 5 6 7')
>       self.assertTrue(res)
E       AssertionError: False is not true

test/units/template/test_templar.py:276: AssertionError
_____________________________________________________________________ TestTemplarIsTemplate.test_is_template_block_undefined ______________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_block_undefined>

    def test_is_template_block_undefined(self):
        res = self.templar.is_template(u'1 2 {%what%} 3 4 {{ foo }} 5 6 7')
>       self.assertTrue(res)
E       AssertionError: False is not true

test/units/template/test_templar.py:272: AssertionError
```

**Review by alikins — CHANGES_REQUESTED** [review]
if the tests failures are behavior as expected / are bad tests, then just the (unintentional?) test/integration changes that need to be fixed. 

And maybe pulling in the test cases.

File: lib/ansible/template/__init__.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/template/__init__.py]: {% what %} is not a valid template but {{foo}} is, it will give a syntax error on templating, not sure on which side this should fall:
a) it has 'tempating markers'
b) its not a valid jinja2 template

File: lib/ansible/template/__init__.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/template/__init__.py]: i think the problem here is our 'typing'

File: lib/ansible/template/__init__.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/template/__init__.py]: we could capture `TempalteSyntaxError` but that would also affect unbalanced `{%`  and similar

File: lib/ansible/template/__init__.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/template/__init__.py]: pushed fix for is_template to handle complex data, before it was just passing to template which always returns either a string or guessed 'typed' result

File: lib/ansible/template/__init__.py [hunk 4]
```diff
@@ -482,6 +499,26 @@ def template(self, variable, convert_bare=False, preserve_trailing_newlines=True
...
+                return True
+            except:
+                return False
+            return (new != data)
+        elif isinstance(data, (list, tuple)):
+            for v in data:
+                if self.is_template(v):
+                    return True
+        elif isinstance(variable, dict):
+            for k in data.:
```
Comment (alikins) [inline lib/ansible/template/__init__.py]: SyntaxError

```
lib/ansible/vars/__init__.py:41: in <module>
    from ansible.template import Templar
E     File "/home/adrian/src/ansible/lib/ansible/template/__init__.py", line 517
E       for k in data.:
E                     ^
E   SyntaxError: invalid syntax
```

**Review by alikins — CHANGES_REQUESTED** [review]
syntax error breaks lots of tests

File: lib/ansible/template/__init__.py [hunk 5]
```diff
@@ -482,6 +499,26 @@ def template(self, variable, convert_bare=False, preserve_trailing_newlines=True
...
+            except UndefinedError:
+                return True
+            except:
+                return False
+            return (new != data)
+        elif isinstance(data, (list, tuple)):
+            for v in data:
+                if self.is_template(v):
+                    return True
+        elif isinstance(variable, dict):
```
Comment (alikins) [inline lib/ansible/template/__init__.py]: undefined var 'variable'

s/variable/data

**Review by alikins — CHANGES_REQUESTED** [review]

**Comment by alikins** [comment]
rebased test_is_instance branch, now down to two test failures

```
________________________________________________________________ TestTemplarIsTemplate.test_is_template_object_with_invalid_jinja _________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_object_with_invalid_jinja>

    def test_is_template_object_with_invalid_jinja(self):
        obj = {'foo': [1, 2, 3, 'bdasdf', '{%what%}', '{{foo}}', 5]}
        res = self.templar.is_template(obj)
>       self.assertFalse(res, 'is_template(%s) return should be False' % obj)
E       AssertionError: is_template({'foo': [1, 2, 3, 'bdasdf', '{%what%}', '{{foo}}', 5]}) return should be False

test/units/template/test_templar.py:291: AssertionError
________________________________________________________________ TestTemplarIsTemplate.test_is_template_tuple_with_unsafe_template ________________________________________________________________
self = <units.template.test_templar.TestTemplarIsTemplate testMethod=test_is_template_tuple_with_unsafe_template>

    def test_is_template_tuple_with_unsafe_template(self):
        some_obj = ('foo', 'bar', wrap_var('{{ some_var }}'))
        res = self.templar.is_template(some_obj)
>       self.assertFalse(res, 'is_template(%s) return should be False' % repr(some_obj))
E       AssertionError: is_template(('foo', 'bar', u'{{ some_var }}')) return should be False

test/units/template/test_templar.py:381: AssertionError
```

**Comment by bcoca** [comment]
Both tests are incorrect as  '{{foo}}' is templatable and an element of the tuple, does not matter the other elements are not templatable. unlike when `{%what%}` is in the same string (making it untemplatable).

**Review by alikins — APPROVED** [review]

**Comment by bcoca** [comment]
backported to 2.3
